File: frame/src/main/java/com/linwei/frame/utils/AppLanguageUtils.java
package com.linwei.frame.utils;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.os.LocaleList;
import android.util.DisplayMetrics;

import java.util.HashMap;
import java.util.Locale;

public class AppLanguageUtils {
    private static String currentLanguage = "";
    public static HashMap<String, Locale> allLanguages = new HashMap<String, Locale>(8) {{
        put("en", Locale.ENGLISH);
        put("zh", Locale.SIMPLIFIED_CHINESE);
        put("in", new Locale("in", "IN", ""));
        put("vi", new Locale("vi", "VN", ""));
    }};

    @SuppressWarnings("deprecation")
    public static void setLanguage(Context context, String language) {
        Resources resources = context.getApplicationContext().getResources();
        DisplayMetrics dm = resources.getDisplayMetrics();
        Configuration config = resources.getConfiguration();

        Locale locale = getLocale(language);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            config.setLocale(locale);
        } else {
            config.locale = locale;
        }
        resources.updateConfiguration(config, dm);
    }

    public static Locale getLocale(String language) {
        if (allLanguages.containsKey(language)) {
            return allLanguages.get(language);
        } else {
            return Locale.getDefault();
        }
    }


    public static Context attachBaseContext(Context context, String language) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return updateResources(context, getLocale(language));
        } else {
            return context;
        }
    }


    @TargetApi(Build.VERSION_CODES.N)
    private static Context updateResources(Context context, Locale locale) {
        Resources resources = context.getResources();

        Configuration configuration = resources.getConfiguration();
        configuration.setLocale(locale);
        configuration.setLocales(new LocaleList(locale));
        return context.createConfigurationContext(configuration);
    }

}

File: cams/src/main/java/com/linwei/cams/utils/AppLanguageUtils.kt
package com.linwei.cams.utils

import android.content.Context
import android.content.res.Configuration
import android.content.res.Resources
import android.os.Build
import android.os.LocaleList
import androidx.annotation.RequiresApi
import com.linwei.cams.ext.isNotNullOrEmpty
import java.util.*

/**
 * ---------------------------------------------------------------------
 * @Author: WeiShuai
 * @Time: 2019/5/19
 * @Contact: linwei9605@gmail.com"
 * @Follow: https://github.com/WeiShuaiDev
 * @Description: 国际化处理
 *-----------------------------------------------------------------------
 */
object AppLanguageUtils {
    /**
     * 系统语言类型
     */
    private val mLanguageTypes: MutableMap<String, Locale> =
        mutableMapOf(
            "zh" to Locale.SIMPLIFIED_CHINESE,
            "en" to Locale.ENGLISH
        )

    /**
     * 增加系统语言类型
     */
    fun addLanguageType(language: String, local: Locale) {
        mLanguageTypes[language] = local
    }

    fun setLanguage(context: Context, language: String) {
        val resources: Resources = context.applicationContext.resources
        val config: Configuration = resources.configuration

        val locale: Locale? = getLocaleByLanguage(language)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            config.setLocale(locale)
        } else {
            @Suppress("DEPRECATION")
            config.locale = locale
        }
        @Suppress("DEPRECATION")
        resources.updateConfiguration(config, resources.displayMetrics)
    }


    fun attachBaseContext(context: Context, language: String): Context {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.N) {
            return updateResources(context, getLocaleByLanguage(language))
        }
        return context
    }

    /**
     * 根据language key获取Locale
     */
    private fun getLocaleByLanguage(language: String?): Locale? {
        if (language.isNotNullOrEmpty()) {
            if (mLanguageTypes.containsKey(language)) {
                return mLanguageTypes[language]
            }
        }
        return Locale.getDefault()
    }

    @RequiresApi(Build.VERSION_CODES.N)
    private fun updateResources(context: Context, locale: Locale?): Context {
        val resources: Resources = context.resources
        val config: Configuration = resources.configuration
        config.setLocale(locale)
        config.setLocales(LocaleList(locale))
        return context.createConfigurationContext(config)
    }
}